File: LeetCodeCPlusPlus/py49.cpp
#include <vector>
#include <string>
#include <sstream>
#include <unordered_map>
using namespace std;
// 字母异位分词
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
         unordered_map<string, vector<string>> hashmap;
         for(auto s : strs){
             vector<int> tmp(26);
             for (auto ch : s)
                 tmp[ch-'a']++;
             std::stringstream ss;
             for (auto k : tmp)
                 ss<<k;
             hashmap[ss.str()].push_back(s);
             //string temp = s;
             //sort(temp.begin(), temp.end());
             //hashmap[temp].push_back(s);
             
         }
        int len = hashmap.size();
        vector<vector<string>> ans(len);
        int index = 0;
        for(auto i : hashmap){
            ans[index++] = i.second;
            //index;
        }
        return ans;

    }
};

File: LeetCodePython/py49.py
# 字母异位分词
# 给定一个字符串数组，将字母异位词组合在一起。字母异位词指字母相同，但排列不同的字符串。
# 解法用26位的散列来作为键值
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        ans = collections.defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1
            ans[tuple(count)].append(s)
        return list(ans.values())